feat(api): add middleware to reject certain endpoint requests if there is no active competition

joewagner opened the pull request:
fixes https://github.com/recallnet/js-recall/issues/985

This was an attempt at an inflight fix for the July 23-24 comp.  It wasn't the root of the issue at the time, but seems like a good performance improvement.

joewagner commented on apps/api/src/middleware/active-comp-filter.middleware.ts:
```diff
@@ -0,0 +1,73 @@
+// Cache for active competition check
+let cachedActiveCompetition: { id: string } | null = null;
+let lastQueryTime = 0;
+const CACHE_ACTIVE_COMP_TTL_MS = config.cache.activeCompetitionTtlMs;
```
Since this is a very short (default 3 seconds) lived cache and a relatively low overhead query, seems fine to just have a per instance in-memory cache that can't be manually reset.  
If we revisit caching in a different layer of the app, we can potentially re-think this.

joewagner commented on apps/api/src/middleware/active-comp-filter.middleware.ts:
```diff
@@ -0,0 +1,71 @@
+        middlewareLogger.debug(
+          "Active comp middleware: No active competition found",
+        );
+        throw new ApiError(403, "No active competition");
```
Simply throw a 403 if there isn't an active competition.  This middleware can be used by any endpoint that is only valid when a competition is happening.

dtbuchholz reviewed (approved):
@joewagner lgmt. but, i noticed some odd behavior that we _could_ choose to fix in this PR. it's sort of related, and maybe we need to write tests for this case. but...it's sort of a separate issue, so we could create a follow up ticket to address the point below.

1. i have an agent trading in an active competition
2. competition ends
3. i try to make a `/trade/execute`
4. instead of this PR's error getting thrown, i get an "invalid api key" error. this is erroneous:
```
{
  "success": false,
  "error": "Invalid API key. This key may have been reset or is no longer associated with an active account. Please ensure you're using your most recent API key."
}
```
e.g., if i make a direct db update and alter the competition's status to `active` and `end_date` to null, i can make the API request in step (1). so, something odd is happening _maybe_ in the other middleware?

i _think_ i've noticed this happen with misc admin API requests, too.

joewagner commented:
> @joewagner lgmt. but, i noticed some odd behavior that we _could_ choose to fix in this PR. it's sort of related, and maybe we need to write tests for this case. but...it's sort of a separate issue, so we could create a follow up ticket to address the point below.
> 
> 1. i have an agent trading in an active competition
> 2. competition ends
> 3. i try to make a `/trade/execute`
> 4. instead of this PR's error getting thrown, i get an "invalid api key" error. this is erroneous:
> 
> ```
> {
>   "success": false,
>   "error": "Invalid API key. This key may have been reset or is no longer associated with an active account. Please ensure you're using your most recent API key."
> }
> ```
> 
> e.g., if i make a direct db update and alter the competition's status to `active` and `end_date` to null, i can make the API request in step (1). so, something odd is happening _maybe_ in the other middleware?
> 
> i _think_ i've noticed this happen with misc admin API requests, too.

Good catch!  Looks like there was a conditional in the auth middleware that looked specifically at the /api/trade/execute method and threw a 401.  I removed that and added two tests.  I also found a problem with the middleware logic, and had to update a bunch of tests to start a comp before getting prices.

This might be worth another look.
